# Sample of Teaching Material

---

I come up with solutions to weekly Python lab assignments for a graduate level introductory data analysis class at the University of San Francisco. The class is composed of college seniors, MBA students, as well as Biotech Masters students who have little or no experience in Python programming. **This is an excerpt from one of my solutions (and a continuation of `teaching_sample_1`)**.

---

### Contents

<br>

<span style="color:black"><b>0.</b> Preliminaries</span>

<a href="#section0_1"><p style="text-indent: 20px"><span style="color:black"><b>0.1.</b> Reading in the data</span></p></a>

<a href="#section0_2"><p style="text-indent: 20px"><span style="color:black"><b>0.2.</b> Exploring the data - <b> \*\*IMPORTANT\*\*</b></span></p></a>

<a href="#section1"><span style="color:black"><b>1.</b> How many guests stayed in total?</span></a>

<a href="#section2"><span style="color:black"><b>2.</b> How many unique property listings are there in this file?</span></a>

<a href="#section3"><span style="color:black"><b>3.</b> Compute the sum total of nights stayed per listing.</span></a>

<a href="#section4"><span style="color:black"><b>4.</b> Compute the sum total of revenue generated per listing.</span></a>

<a href="#section5"><span style="color:black"><b>5.</b> Compute the mean revenue generated per night per listing.</span></a>

<a href="#section6"><span style="color:black"><b>6.</b> Compute the mean length of stay per listing.</span></a>

<a href="#section7"><span style="color:black"><b>7.</b> How many stays were shorter than 3 days?</span></a>

<a href="#section8"><span style="color:black"><b>8.</b> How many stays exceeded 5 days?</span></a>

<a href="#section9"><span style="color:black"><b>9.</b> How many stays included one weekend day?</span></a>

<a href="#section10"><span style="color:black"><b>10.</b> How many stays began on a Tuesday?</span></a>

<a href="#section11"><span style="color:black"><b>11.</b> What is the sum total commission paid to Airbnb by listing?</span></a>

<a href="#section12"><span style="color:black"><b>12.</b> What is the sum total of taxes paid to Airbnb?</span></a>


---

<em>Fun Fact:</em> Notice how this contents page is very similar to a Python dictionary? By having these "keys" that lead you straight to the solution for the particular question you are interested in, you don't have to scroll painfully through this document. **A lot of time is saved!** :)

---

<a id='section0_1'></a>
<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        Read in <code>airbnb.csv</code>.
    </span>
</div>

In [160]:
## Import Pandas library.
import pandas as pd

In [161]:
## NOTE: Put airbnb.csv in the same folder as your notebook for the code below to work (and for your own convenience).
df = pd.read_csv("airbnb.csv")

-----

<a id='section0_2'></a>
<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        Explore <code>airbnb.csv</code>.
    </span>
</div>

---
> **Note:** This is the section where the sheer importance of <u>Exploratory Data Analysis (EDA)</u> comes to light. Before you can answer complex questions with and fit <u>appropriate</u> machine learning models to your data, you need to <u>understand your data well enough</u>!

---

Let's look at the dataset:

In [162]:
df

Date         Type Confirmation Code  Start Date  Nights  \
0    08/23/2019       Payout               NaN         NaN     NaN   
1    08/23/2019  Reservation        HMANZTA9TH  08/22/2019     3.0   
2    08/21/2019       Payout               NaN         NaN     NaN   
3    08/21/2019  Reservation        HMADNHAT3X  08/20/2019     2.0   
4    08/18/2019       Payout               NaN         NaN     NaN   
..          ...          ...               ...         ...     ...   
119  01/12/2019  Reservation        HMS3XFNTC9  01/10/2019     3.0   
120  01/10/2019       Payout               NaN         NaN     NaN   
121  01/10/2019  Reservation        HMEY8DKYEN  01/08/2019     2.0   
122  01/08/2019       Payout               NaN         NaN     NaN   
123  01/08/2019  Reservation        HM2HMWYKXZ  01/06/2019     2.0   

                 Guest                                            Listing  \
0                  NaN                                                NaN   
1    Katelyn Lindstrom  Modern Gem with Upscale Finishes in San Francisco   
2                  NaN                                                NaN   
3        Erik Engström  Modern Gem with Upscale Finishes in San Francisco   
4                  NaN                                                NaN   
..                 ...                                                ...   
119             D Ayde  Modern Gem with Upscale Finishes in San Francisco   
120                NaN                                                NaN   
121     Shahab Kashani  Modern Gem with Upscale Finishes in San Francisco   
122                NaN                                                NaN   
123      Jerome Pascua  Modern Gem with Upscale Finishes in San Francisco   

                         Details  Reference Currency  Amount  Paid Out  \
0    Transfer to •••••7473 (USD)        NaN      USD     NaN    446.20   
1                            NaN        NaN      USD  446.20       NaN   
2    Transfer to •••••7473 (USD)        NaN      USD     NaN    310.40   
3                            NaN        NaN      USD  310.40       NaN   
4    Transfer to •••••7473 (USD)        NaN      USD     NaN    431.65   
..                           ...        ...      ...     ...       ...   
119                          NaN        NaN      USD  341.44       NaN   
120  Transfer to •••••7473 (USD)        NaN      USD     NaN    235.71   
121                          NaN        NaN      USD  235.71       NaN   
122  Transfer to •••••7473 (USD)        NaN      USD     NaN    245.41   
123                          NaN        NaN      USD  245.41       NaN   

     Host Fee  Cleaning Fee  
0         NaN           NaN  
1       13.80          70.0  
2         NaN           NaN  
3        9.60          70.0  
4         NaN           NaN  
..        ...           ...  
119     10.56          65.0  
120       NaN           NaN  
121      7.29          65.0  
122       NaN           NaN  
123      7.59          65.0  

[124 rows x 14 columns]

The dataset looks weird (and therefore, potentially annoying). There seem to be alternating rows with *lots* of `NaN` values!


Seeing how 'Payout', 'Reservation', and financial numbers appear frequently, we can surmise that the dataset involves Airbnb\* transactions. But are these transactions between hosts, guests, Airbnb, or what? 

> \* This we know from the name of the dataset (`airbnb.csv`).


What does 'Payout' even mean? According to [Airbnb](https://www.airbnb.com/resources/hosting-homes/a/how-you-get-paid-for-hosting-20), 'Payout' seems to be money that Airbnb receives from a guest and transmits to a host. 

> So: Airbnb is like a middleman for financial transactions here. This is consistent with what we know about Airbnb, that it is a C2C platform.

This suggests that the dataset might simply contain data of transactions between Airbnb and hosts... or *a* host? It appears that payouts are being transmitted to a particular bank account number (•••••7473). To be certain, let's look at the distribution of values in the **Details** column:

In [163]:
df['Details'].value_counts()

Transfer to •••••7473 (USD)                           58
Resolution adjustment for resolution 12100879          1
Photography payment for listing 24357954 at Feb 23     1
Resolution adjustment for resolution 11976696          1
Resolution payout for resolution 13044609              1
Name: Details, dtype: int64

So, it does seem like most of the records in the dataset involve payments to that *one* particular bank account number.


We can also look at the distribution of values in the **Listing** column:

In [164]:
df['Listing'].value_counts()

Modern Gem with Upscale Finishes in San Francisco    51
Charming Family Home in San Francisco                10
Charming San Francisco Edwardian w/ Private Suite     1
Name: Listing, dtype: int64

Seems like there are only three properties involved in this dataset. 


These two pieces of information strongly suggest that the dataset is a collection of transaction details between Airbnb and *a particular host*. 

<a id='date_range'></a>
Time-stamped from January 2019 to August 2019 (see below).

In [165]:
print("Earliest Date: ", df['Date'].min())
print("Latest Date: ", df['Date'].max())

Earliest Date:  01/08/2019
Latest Date:  08/23/2019


Looking at the data again, it also *seems* that every row with a non-`NaN` value under the **Amount** column AND 'Reservation' under the **Type** column is immediately preceded by a row with an <u>exactly matching value</u> under the **Paid Out** column AND 'Payout' under the **Type** column.


This suggests that the data might be a sequence of pairs of adjacent records of a guest's reservation and the payment associated with the reservation.


But is it *just* composed of such pairs of adjacent records? Let's check.

In [166]:
## High-level check: Get a sense of the distribution of values in the "Type" column.
df["Type"].value_counts()

Reservation              62
Payout                   58
Resolution Adjustment     2
Resolution Payout         1
Photography Payment       1
Name: Type, dtype: int64

So, it does seem that *most* of the records *may be* pairs\* of adjacent records.

> \* We will need to do a detailed check later to be sure that those 58 Payout records are *in fact* paired with 58 of the 62 Reservation records. (We will see how to perform this check specifically, later.)

What are those four records $-$ 2 for Resolution Adjustment, 1 for Resolution Payout, and 1 for Photography Payment $-$ doing in the data? Let's see.

In [167]:
## Investigate those four interesting records.
df.loc[(df['Type'] != 'Reservation') & (df['Type'] != 'Payout')]

Date                   Type Confirmation Code Start Date  Nights  \
29  06/22/2019      Resolution Payout               NaN        NaN     NaN   
62  04/29/2019  Resolution Adjustment               NaN        NaN     NaN   
70  04/22/2019  Resolution Adjustment               NaN        NaN     NaN   
93  02/26/2019    Photography Payment               NaN        NaN     NaN   

   Guest Listing                                            Details  \
29   NaN     NaN          Resolution payout for resolution 13044609   
62   NaN     NaN      Resolution adjustment for resolution 12100879   
70   NaN     NaN      Resolution adjustment for resolution 11976696   
93   NaN     NaN  Photography payment for listing 24357954 at Fe...   

    Reference Currency  Amount  Paid Out  Host Fee  Cleaning Fee  
29        NaN      USD   145.5       NaN       NaN           NaN  
62        NaN      USD   -50.0       NaN       NaN           NaN  
70        NaN      USD   -50.0       NaN       NaN           NaN  
93        NaN      USD  -147.0       NaN       NaN           NaN

The four records look interesting indeed: the last three have <u>negative</u> values in the **Amount** column. Does this mean that for these three records $-$ 2 involving Resolution Adjustments and 1 involving Photography Payment $-$ the host actually paid the specified amounts to Airbnb? 

---
<em>A short digression</em>

Let's see if there are other records that also have negative values in the **Amount** column. 

>If there are, that could complicate our analysis later...

In [168]:
## Investigate records with negative Amount values.
df.loc[df['Amount'] < 0]

Date                   Type Confirmation Code Start Date  Nights  \
62  04/29/2019  Resolution Adjustment               NaN        NaN     NaN   
70  04/22/2019  Resolution Adjustment               NaN        NaN     NaN   
93  02/26/2019    Photography Payment               NaN        NaN     NaN   

   Guest Listing                                            Details  \
62   NaN     NaN      Resolution adjustment for resolution 12100879   
70   NaN     NaN      Resolution adjustment for resolution 11976696   
93   NaN     NaN  Photography payment for listing 24357954 at Fe...   

    Reference Currency  Amount  Paid Out  Host Fee  Cleaning Fee  
62        NaN      USD   -50.0       NaN       NaN           NaN  
70        NaN      USD   -50.0       NaN       NaN           NaN  
93        NaN      USD  -147.0       NaN       NaN           NaN

Looks like these three records are *identical* to the three records we saw earlier that had negative values in the **Amount** column. How lucky we are!

---

*Back to the four interesting records*

If we do some research, we find that a [Resolution Adjustment](https://www.airbnb.com/help/article/459/calculating-your-payout) might happen when the host refunds a guest. Also, a [Photography Payment](https://community.withairbnb.com/t5/Help/superhost-bonus/td-p/1048295) seems to be a fee that Airbnb charged the host for helping him/her take photographs of the property.


So yes, it does seem like with regard to the three records, the host had to pay either Airbnb or a guest (via Airbnb).


For the single record involving a Resolution Payout, relevant information could not be found on the web. But intuition suggests that this might happen when a guest damages the host's property and has to pay for the damages, or when Airbnb has to pay the host for whatever reason.

---

Another thing we could do to obtain more useful contextual information about the four interesting records is to look at their *neigbouring records*.

<a id='1st_interesting_record'></a>
**1st interesting record**

In [169]:
df.iloc[29:30]

Date               Type Confirmation Code Start Date  Nights Guest  \
29  06/22/2019  Resolution Payout               NaN        NaN     NaN   NaN   

   Listing                                    Details  Reference Currency  \
29     NaN  Resolution payout for resolution 13044609        NaN      USD   

    Amount  Paid Out  Host Fee  Cleaning Fee  
29   145.5       NaN       NaN           NaN

Let's look at the records in the vicinity of row **29** (above). 


A cursory glance (+ some quick mental math) suggests that rows **28** to **31** might actually be related $-$ the numbers in the **Amount** column for rows **29** to **31** seem to add up to the number in the **Paid Out** column for row **28**!

In [170]:
df.iloc[26:36] # The numbers in rows 28-31 actually match up!

Date               Type Confirmation Code  Start Date  Nights  \
26  06/27/2019             Payout               NaN         NaN     NaN   
27  06/27/2019        Reservation        HMSZQWB83R  06/26/2019     4.0   
28  06/22/2019             Payout               NaN         NaN     NaN   
29  06/22/2019  Resolution Payout               NaN         NaN     NaN   
30  06/22/2019        Reservation        HMAAEB8BRK  06/21/2019     2.0   
31  06/22/2019        Reservation        HMAQW2PHXY  06/21/2019     1.0   
32  06/19/2019             Payout               NaN         NaN     NaN   
33  06/19/2019        Reservation        HMANEJTK4E  06/18/2019     3.0   
34  06/15/2019             Payout               NaN         NaN     NaN   
35  06/15/2019        Reservation        HMAN4N55ZJ  06/14/2019     4.0   

               Guest                                            Listing  \
26               NaN                                                NaN   
27  Renato Rey Perez  Charming San Francisco Edwardian w/ Private Suite   
28               NaN                                                NaN   
29               NaN                                                NaN   
30        Alec Cohen  Modern Gem with Upscale Finishes in San Francisco   
31       Maddy Smith  Modern Gem with Upscale Finishes in San Francisco   
32               NaN                                                NaN   
33     Lianne Aratea  Modern Gem with Upscale Finishes in San Francisco   
34               NaN                                                NaN   
35    Robert Yacucci  Modern Gem with Upscale Finishes in San Francisco   

                                      Details  Reference Currency   Amount  \
26                Transfer to •••••7473 (USD)        NaN      USD      NaN   
27                                        NaN        NaN      USD  1742.12   
28                Transfer to •••••7473 (USD)        NaN      USD      NaN   
29  Resolution payout for resolution 13044609        NaN      USD   145.50   
30                                        NaN        NaN      USD   115.43   
31                                        NaN        NaN      USD   213.40   
32                Transfer to •••••7473 (USD)        NaN      USD      NaN   
33                                        NaN        NaN      USD   385.09   
34                Transfer to •••••7473 (USD)        NaN      USD      NaN   
35                                        NaN        NaN      USD   510.22   

    Paid Out  Host Fee  Cleaning Fee  
26   1742.12       NaN           NaN  
27       NaN     53.88         200.0  
28    474.33       NaN           NaN  
29       NaN       NaN           NaN  
30       NaN      1.38           0.0  
31       NaN      6.60          70.0  
32    385.09       NaN           NaN  
33       NaN     11.91          70.0  
34    510.22       NaN           NaN  
35       NaN     15.78          70.0

We can do some calculations to be sure:

In [171]:
## Check that numbers in Amount column for rows 29 to 31 actually add up to number in Paid Out column for row 28.
num_28 = df.loc[28, 'Paid Out']
num_28 = round(num_28, 2)
print("Number in Paid Out column for row 28: ", num_28)
print()
num_29_31 = df.loc[29:31, 'Amount'].sum()
num_29_31 = round(num_29_31, 2)
print("Sum of numbers in Amount column for rows 29 to 31: ", num_29_31)
print()
print("We were right: ", num_28 == num_29_31)

Number in Paid Out column for row 28:  474.33

Sum of numbers in Amount column for rows 29 to 31:  474.33

We were right:  True


---

**2nd interesting record**

In [172]:
df.iloc[62:63]

Date                   Type Confirmation Code Start Date  Nights  \
62  04/29/2019  Resolution Adjustment               NaN        NaN     NaN   

   Guest Listing                                        Details  Reference  \
62   NaN     NaN  Resolution adjustment for resolution 12100879        NaN   

   Currency  Amount  Paid Out  Host Fee  Cleaning Fee  
62      USD   -50.0       NaN       NaN           NaN

Let's look at the records in the vicinity of row **62** (see above). 


A cursory glance (+ some quick mental math) suggests that rows **61** to **64** might actually be related $-$ the numbers in the **Amount** column for rows **62** to **64** seem to add up to the number in the **Paid Out** column for row **61**!

In [173]:
df.iloc[59:67] # The numbers in rows 61-64 actually match up!

Date                   Type Confirmation Code  Start Date  Nights  \
59  05/04/2019                 Payout               NaN         NaN     NaN   
60  05/04/2019            Reservation        HMAEARM8CP  05/03/2019     2.0   
61  05/01/2019                 Payout               NaN         NaN     NaN   
62  04/29/2019  Resolution Adjustment               NaN         NaN     NaN   
63  05/01/2019            Reservation        HMYQNM2ECW  04/30/2019     5.0   
64  05/01/2019            Reservation        HMAKW2PEZW  04/30/2019     3.0   
65  04/27/2019                 Payout               NaN         NaN     NaN   
66  04/27/2019            Reservation        HMARRKJPA8  04/26/2019     4.0   

             Guest                                            Listing  \
59             NaN                                                NaN   
60   Jules Vasquez  Modern Gem with Upscale Finishes in San Francisco   
61             NaN                                                NaN   
62             NaN                                                NaN   
63  Lauren Provini  Modern Gem with Upscale Finishes in San Francisco   
64  John Bashmakov  Modern Gem with Upscale Finishes in San Francisco   
65             NaN                                                NaN   
66   Sunny Burbank  Modern Gem with Upscale Finishes in San Francisco   

                                          Details  Reference Currency  Amount  \
59                    Transfer to •••••7473 (USD)        NaN      USD     NaN   
60                                            NaN        NaN      USD  298.76   
61                    Transfer to •••••7473 (USD)        NaN      USD     NaN   
62  Resolution adjustment for resolution 12100879        NaN      USD  -50.00   
63                                            NaN        NaN      USD  274.02   
64                                            NaN        NaN      USD  385.09   
65                    Transfer to •••••7473 (USD)        NaN      USD     NaN   
66                                            NaN        NaN      USD  510.22   

    Paid Out  Host Fee  Cleaning Fee  
59    298.76       NaN           NaN  
60       NaN      9.24          70.0  
61    609.11       NaN           NaN  
62       NaN       NaN           NaN  
63       NaN      3.80           0.0  
64       NaN     11.91          70.0  
65    510.22       NaN           NaN  
66       NaN     15.78          70.0

We can do some calculations to be sure:

In [174]:
## Check that numbers in Amount column for rows 62 to 64 actually add up to number in Paid Out column for row 61.
num_61 = df.loc[61, 'Paid Out']
num_61 = round(num_61, 2)
print("Number in Paid Out column for row 61: ", num_61)
print()
num_62_64 = df.loc[62:64, 'Amount'].sum()
num_62_64 = round(num_62_64, 2)
print("Sum of numbers in Amount column for rows 62 to 64: ", num_62_64)
print()
print("We were right: ", num_61 == num_62_64)

Number in Paid Out column for row 61:  609.11

Sum of numbers in Amount column for rows 62 to 64:  609.11

We were right:  True


---

**3rd interesting record**

In [175]:
df.iloc[70:71]

Date                   Type Confirmation Code Start Date  Nights  \
70  04/22/2019  Resolution Adjustment               NaN        NaN     NaN   

   Guest Listing                                        Details  Reference  \
70   NaN     NaN  Resolution adjustment for resolution 11976696        NaN   

   Currency  Amount  Paid Out  Host Fee  Cleaning Fee  
70      USD   -50.0       NaN       NaN           NaN

Let's look at the records in the vicinity of row **70** (see above). 


A cursory glance (+ some quick mental math) suggests that rows **69** to **71** might actually be related $-$ the numbers in the **Amount** column for rows **70** to **71** seem to add up to the number in the **Paid Out** column for row **69**!

In [176]:
df.iloc[67:74] # The numbers in rows 69-71 actually match up!

Date                   Type Confirmation Code  Start Date  Nights  \
67  04/23/2019                 Payout               NaN         NaN     NaN   
68  04/23/2019            Reservation        HMAEKB52PQ  04/22/2019     3.0   
69  04/23/2019                 Payout               NaN         NaN     NaN   
70  04/22/2019  Resolution Adjustment               NaN         NaN     NaN   
71  04/22/2019            Reservation        HMAKS82CT4  04/21/2019     1.0   
72  04/13/2019                 Payout               NaN         NaN     NaN   
73  04/13/2019            Reservation        HMCF5FDBWK  04/12/2019     2.0   

                Guest                                            Listing  \
67                NaN                                                NaN   
68  Katharina Volkmar  Modern Gem with Upscale Finishes in San Francisco   
69                NaN                                                NaN   
70                NaN                                                NaN   
71  Madelyn Wentworth  Modern Gem with Upscale Finishes in San Francisco   
72                NaN                                                NaN   
73         Radu Butoi  Modern Gem with Upscale Finishes in San Francisco   

                                          Details  Reference Currency  Amount  \
67                    Transfer to •••••7473 (USD)        NaN      USD     NaN   
68                                            NaN        NaN      USD  385.09   
69                    Transfer to •••••7473 (USD)        NaN      USD     NaN   
70  Resolution adjustment for resolution 11976696        NaN      USD  -50.00   
71                                            NaN        NaN      USD  173.63   
72                    Transfer to •••••7473 (USD)        NaN      USD     NaN   
73                                            NaN        NaN      USD  293.91   

    Paid Out  Host Fee  Cleaning Fee  
67    385.09       NaN           NaN  
68       NaN     11.91          70.0  
69    123.63       NaN           NaN  
70       NaN       NaN           NaN  
71       NaN      5.37          70.0  
72    293.91       NaN           NaN  
73       NaN      9.09          65.0

We can do some calculations to be sure:

In [177]:
## Check that numbers in Amount column for rows 70 to 71 actually add up to number in Paid Out column for row 69.
num_69 = df.loc[69, 'Paid Out']
num_69 = round(num_69, 2)
print("Number in Paid Out column for row 69: ", num_69)
print()
num_70_71 = df.loc[70:71, 'Amount'].sum()
num_70_71 = round(num_70_71, 2)
print("Sum of numbers in Amount column for rows 70 to 71: ", num_70_71)
print()
print("We were right: ", num_69 == num_70_71)

Number in Paid Out column for row 69:  123.63

Sum of numbers in Amount column for rows 70 to 71:  123.63

We were right:  True


---

**4th (and final!) interesting record**

In [178]:
df.iloc[93:94]

Date                 Type Confirmation Code Start Date  Nights  \
93  02/26/2019  Photography Payment               NaN        NaN     NaN   

   Guest Listing                                            Details  \
93   NaN     NaN  Photography payment for listing 24357954 at Fe...   

    Reference Currency  Amount  Paid Out  Host Fee  Cleaning Fee  
93        NaN      USD  -147.0       NaN       NaN           NaN

Let's look at the records in the vicinity of row **93** (see above). 


A cursory glance (+ some quick mental math) suggests that rows **92** to **95** might actually be related $-$ the numbers in the **Amount** column for rows **93** to **95** seem to add up to the number in the **Paid Out** column for row **92**!

In [179]:
df.iloc[90:98] # The numbers in rows 92-95 actually match up!

Date                 Type Confirmation Code  Start Date  Nights  \
90  03/06/2019               Payout               NaN         NaN     NaN   
91  03/07/2019          Reservation        HMMWNJPF4B  03/05/2019     2.0   
92  03/03/2019               Payout               NaN         NaN     NaN   
93  02/26/2019  Photography Payment               NaN         NaN     NaN   
94  03/02/2019          Reservation        HMEEP8EA9D  02/28/2019     2.0   
95  03/03/2019          Reservation        HMXHTWYYS8  03/01/2019     4.0   
96  02/24/2019               Payout               NaN         NaN     NaN   
97  02/25/2019          Reservation        HMX4YQNFN9  02/23/2019     3.0   

                  Guest                                            Listing  \
90                  NaN                                                NaN   
91  Christopher Rodgers  Modern Gem with Upscale Finishes in San Francisco   
92                  NaN                                                NaN   
93                  NaN                                                NaN   
94          Lucy Nguyen  Modern Gem with Upscale Finishes in San Francisco   
95                 明月 李  Modern Gem with Upscale Finishes in San Francisco   
96                  NaN                                                NaN   
97         Ella Stimson  Modern Gem with Upscale Finishes in San Francisco   

                                              Details  Reference Currency  \
90                        Transfer to •••••7473 (USD)        NaN      USD   
91                                                NaN        NaN      USD   
92                        Transfer to •••••7473 (USD)        NaN      USD   
93  Photography payment for listing 24357954 at Fe...        NaN      USD   
94                                                NaN        NaN      USD   
95                                                NaN        NaN      USD   
96                        Transfer to •••••7473 (USD)        NaN      USD   
97                                                NaN        NaN      USD   

    Amount  Paid Out  Host Fee  Cleaning Fee  
90     NaN    255.11       NaN           NaN  
91  255.11       NaN      7.89          65.0  
92     NaN    420.45       NaN           NaN  
93 -147.00       NaN       NaN           NaN  
94  100.88       NaN      1.19           0.0  
95  466.57       NaN     14.43          65.0  
96     NaN    360.84       NaN           NaN  
97  360.84       NaN     11.16          65.0

We can do some calculations to be sure:

In [180]:
## Check that numbers in Amount column for rows 93 to 95 actually add up to number in Paid Out column for row 92.
num_92 = df.loc[92, 'Paid Out']
num_92 = round(num_92, 2)
print("Number in Paid Out column for row 92: ", num_92)
print()
num_93_95 = df.loc[93:95, 'Amount'].sum()
num_93_95 = round(num_93_95, 2)
print("Sum of numbers in Amount column for rows 93 to 95: ", num_93_95)
print()
print("We were right: ", num_92 == num_93_95)

Number in Paid Out column for row 92:  420.45

Sum of numbers in Amount column for rows 93 to 95:  420.45

We were right:  True


---

So, it seems these four interesting records are basically of non-standard transactions <em>in relation to existing reservations</em>. 


Further the fact that, for each group (containing an interesting record),

<br>

$$\text{Reservation(s) Amount}\hspace{2mm} +\hspace{2mm} \text{Non-standard transaction(s) Amount}\hspace{2mm} =\hspace{2mm} \text{Paid Out Amount}$$

<br>

means that every reservation (or set of reservations) has been paid for in the case of the four groups.


What about the remaining records outside of the four groups? Are they *truly* all just pairs of Reservation-Payout records? It is still difficult to say anything conclusive because, as it turns out, the number of Reservation records don't actually match the number of Payout records! (See below for proof.)

---
*Proof that among the remaining records, # of Reservation records $\neq$ # of Payout records*

Recall the distribution of values in the **Type** column for the entire dataset:

In [181]:
df["Type"].value_counts()

Reservation              62
Payout                   58
Resolution Adjustment     2
Resolution Payout         1
Photography Payment       1
Name: Type, dtype: int64

Now, if we count the number of Reservation and Payout records in each of the four groups, we will get the following:

Group | # of Reservation records | # of Payout records
:-:|:-:|:-:
Rows 28 to 31 | 2 | 1
Rows 61 to 64 | 2 | 1
Rows 69 to 71 | 1 | 1
Rows 92 to 95 | 2 | 1
*Total* | 7 | 4

If we do the math of excluding the four groups from the dataset, then there would be 62 - 7 = 55 Reservation records and 58 - 4 = 54 Payout records.


This means the number of Reservation records is 1 more than than the number of Payout records! What could explain this *mismatch*? 


Could it be that there is a guest who made a reservation but did not follow through with the payment\*?

> \* which might imply that this guest did *not* follow through with the reservation and stay at the place. (Big if true; consider [**Question 1**](#section1).)

Or, could it be that there is a set of *two* reservations for a particular payout?


Let's see:

In [182]:
## Store records from the overall data that are NOT in the four groups in a new DataFrame.

normal_df = pd.concat([df.loc[:27], df.loc[32:60], df.loc[65:68], df.loc[72:91]], axis=0).reset_index(drop=True)
normal_df

Date         Type Confirmation Code  Start Date  Nights  \
0   08/23/2019       Payout               NaN         NaN     NaN   
1   08/23/2019  Reservation        HMANZTA9TH  08/22/2019     3.0   
2   08/21/2019       Payout               NaN         NaN     NaN   
3   08/21/2019  Reservation        HMADNHAT3X  08/20/2019     2.0   
4   08/18/2019       Payout               NaN         NaN     NaN   
..         ...          ...               ...         ...     ...   
76  03/12/2019  Reservation        HMAHD9QBH3  03/10/2019     4.0   
77  03/08/2019       Payout               NaN         NaN     NaN   
78  03/09/2019  Reservation        HMXTE3F2SB  03/07/2019     3.0   
79  03/06/2019       Payout               NaN         NaN     NaN   
80  03/07/2019  Reservation        HMMWNJPF4B  03/05/2019     2.0   

                  Guest                                            Listing  \
0                   NaN                                                NaN   
1     Katelyn Lindstrom  Modern Gem with Upscale Finishes in San Francisco   
2                   NaN                                                NaN   
3         Erik Engström  Modern Gem with Upscale Finishes in San Francisco   
4                   NaN                                                NaN   
..                  ...                                                ...   
76         Ashlie Flinn  Modern Gem with Upscale Finishes in San Francisco   
77                  NaN                                                NaN   
78      Jessica Soliven  Modern Gem with Upscale Finishes in San Francisco   
79                  NaN                                                NaN   
80  Christopher Rodgers  Modern Gem with Upscale Finishes in San Francisco   

                        Details  Reference Currency  Amount  Paid Out  \
0   Transfer to •••••7473 (USD)        NaN      USD     NaN    446.20   
1                           NaN        NaN      USD  446.20       NaN   
2   Transfer to •••••7473 (USD)        NaN      USD     NaN    310.40   
3                           NaN        NaN      USD  310.40       NaN   
4   Transfer to •••••7473 (USD)        NaN      USD     NaN    431.65   
..                          ...        ...      ...     ...       ...   
76                          NaN        NaN      USD  413.22       NaN   
77  Transfer to •••••7473 (USD)        NaN      USD     NaN    370.54   
78                          NaN        NaN      USD  370.54       NaN   
79  Transfer to •••••7473 (USD)        NaN      USD     NaN    255.11   
80                          NaN        NaN      USD  255.11       NaN   

    Host Fee  Cleaning Fee  
0        NaN           NaN  
1      13.80          70.0  
2        NaN           NaN  
3       9.60          70.0  
4        NaN           NaN  
..       ...           ...  
76     12.78          70.0  
77       NaN           NaN  
78     11.46          65.0  
79       NaN           NaN  
80      7.89          65.0  

[81 rows x 14 columns]

In [183]:
## Find the cause of the inconsistency: if any Reservation record doesn't pair with preceding Payout record, flag it.
idx = 0
no_mismatch = True
while no_mismatch:
    if normal_df['Paid Out'].tolist()[:-1][idx] != normal_df['Amount'].tolist()[1:][idx]:
        no_mismatch = False
    else:
        idx += 2

## Look at the neigbouring records of the anomalous record detected.
normal_df.loc[idx-2: idx+4]

Date         Type Confirmation Code  Start Date  Nights  \
36  06/06/2019       Payout               NaN         NaN     NaN   
37  06/06/2019  Reservation        HMZMDEAJWP  06/05/2019     3.0   
38  06/02/2019       Payout               NaN         NaN     NaN   
39  05/29/2019  Reservation        HMBDZ2PBFC  05/28/2019     4.0   
40  06/02/2019  Reservation        HMW4EBB2RS  06/01/2019     3.0   
41  05/26/2019       Payout               NaN         NaN     NaN   
42  05/26/2019  Reservation        HMAFC4KRTR  05/25/2019     3.0   

                 Guest                                            Listing  \
36                 NaN                                                NaN   
37          Kelly Weir  Modern Gem with Upscale Finishes in San Francisco   
38                 NaN                                                NaN   
39      Ashley Leonard  Modern Gem with Upscale Finishes in San Francisco   
40  Christoph Thumfart  Modern Gem with Upscale Finishes in San Francisco   
41                 NaN                                                NaN   
42                소민 김  Modern Gem with Upscale Finishes in San Francisco   

                        Details  Reference Currency  Amount  Paid Out  \
36  Transfer to •••••7473 (USD)        NaN      USD     NaN    389.94   
37                          NaN        NaN      USD  389.94       NaN   
38  Transfer to •••••7473 (USD)        NaN      USD     NaN    880.76   
39                          NaN        NaN      USD  485.97       NaN   
40                          NaN        NaN      USD  394.79       NaN   
41  Transfer to •••••7473 (USD)        NaN      USD     NaN    375.39   
42                          NaN        NaN      USD  375.39       NaN   

    Host Fee  Cleaning Fee  
36       NaN           NaN  
37     12.06          65.0  
38       NaN           NaN  
39     15.03          65.0  
40     12.21          70.0  
41       NaN           NaN  
42     11.61          70.0

It appears that the cause of the anomaly is that there are two consecutive Reservation records from rows **39** to **40**. In fact, some quick mental math suggests that the numbers in the **Amount** column for rows **39** to **40** seem to add up to the number in the **Paid Out** column for row **38**!


We can do some calculations to be sure:

In [184]:
## Check that numbers in Amount column for rows 39 to 40 actually add up to number in Paid Out column for row 38.
num_38 = normal_df.loc[38, 'Paid Out']
num_38 = round(num_38, 2)
print("Number in Paid Out column for row 38: ", num_38)
print()
num_39_40 = normal_df.loc[39:40, 'Amount'].sum()
num_39_40 = round(num_39_40, 2)
print("Sum of numbers in Amount column for rows 39 to 40: ", num_39_40)
print()
print("We were right: ", num_38 == num_39_40)

Number in Paid Out column for row 38:  880.76

Sum of numbers in Amount column for rows 39 to 40:  880.76

We were right:  True


So now we know the cause of the anomaly: there is a set of *two* reservations for a particular payout. 


This means that up till now, we still have not found any record of an *unpaid* reservation. Can we prove definitively that <em>all</em> reservations have been paid for? 


Let's see.

In [185]:
## Filter out the group containing the anomalous records.
normal_df = pd.concat([normal_df.loc[:37], normal_df.loc[41:]], axis=0).reset_index(drop=True)
normal_df

Date         Type Confirmation Code  Start Date  Nights  \
0   08/23/2019       Payout               NaN         NaN     NaN   
1   08/23/2019  Reservation        HMANZTA9TH  08/22/2019     3.0   
2   08/21/2019       Payout               NaN         NaN     NaN   
3   08/21/2019  Reservation        HMADNHAT3X  08/20/2019     2.0   
4   08/18/2019       Payout               NaN         NaN     NaN   
..         ...          ...               ...         ...     ...   
73  03/12/2019  Reservation        HMAHD9QBH3  03/10/2019     4.0   
74  03/08/2019       Payout               NaN         NaN     NaN   
75  03/09/2019  Reservation        HMXTE3F2SB  03/07/2019     3.0   
76  03/06/2019       Payout               NaN         NaN     NaN   
77  03/07/2019  Reservation        HMMWNJPF4B  03/05/2019     2.0   

                  Guest                                            Listing  \
0                   NaN                                                NaN   
1     Katelyn Lindstrom  Modern Gem with Upscale Finishes in San Francisco   
2                   NaN                                                NaN   
3         Erik Engström  Modern Gem with Upscale Finishes in San Francisco   
4                   NaN                                                NaN   
..                  ...                                                ...   
73         Ashlie Flinn  Modern Gem with Upscale Finishes in San Francisco   
74                  NaN                                                NaN   
75      Jessica Soliven  Modern Gem with Upscale Finishes in San Francisco   
76                  NaN                                                NaN   
77  Christopher Rodgers  Modern Gem with Upscale Finishes in San Francisco   

                        Details  Reference Currency  Amount  Paid Out  \
0   Transfer to •••••7473 (USD)        NaN      USD     NaN    446.20   
1                           NaN        NaN      USD  446.20       NaN   
2   Transfer to •••••7473 (USD)        NaN      USD     NaN    310.40   
3                           NaN        NaN      USD  310.40       NaN   
4   Transfer to •••••7473 (USD)        NaN      USD     NaN    431.65   
..                          ...        ...      ...     ...       ...   
73                          NaN        NaN      USD  413.22       NaN   
74  Transfer to •••••7473 (USD)        NaN      USD     NaN    370.54   
75                          NaN        NaN      USD  370.54       NaN   
76  Transfer to •••••7473 (USD)        NaN      USD     NaN    255.11   
77                          NaN        NaN      USD  255.11       NaN   

    Host Fee  Cleaning Fee  
0        NaN           NaN  
1      13.80          70.0  
2        NaN           NaN  
3       9.60          70.0  
4        NaN           NaN  
..       ...           ...  
73     12.78          70.0  
74       NaN           NaN  
75     11.46          65.0  
76       NaN           NaN  
77      7.89          65.0  

[78 rows x 14 columns]

In [186]:
## Check for additional anomalies in further-filtered data.
idx = 0
no_mismatch = True
while no_mismatch:
    if idx >= len(normal_df['Paid Out'].tolist()[:-1]):
        print("No anomalies detected. Fantastic!")
        break
    elif normal_df['Paid Out'].tolist()[:-1][idx] != normal_df['Amount'].tolist()[1:][idx]:
        no_mismatch = False
        print("Another anomaly detected!")
    else:
        idx += 2

No anomalies detected. Fantastic!


It seems that the remaining data are really just pairs of Reservation-Payout records. 


This implies that *every* reservation has been paid for.

<a id='section1'></a>
### Question 1

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        How many guests stayed in total?
    </span>
</div>

To answer this question, we need to know when a guest has stayed. 

> This may be trickier than you think. Note that the question isn't just asking how many guests there are in the dataset, but how many <em>stayed</em>!

Given what we now understand about the data, the best criterion might simply be this: 

> A guest has stayed iff he/she has not only <u>made a reservation</u> but also <u>paid for the reservation</u>.


Since we proved earlier that all reservations in the data have been paid for, we must conclude that every guest in the dataset *is* a guest who stayed. Thus,

<br>

$$\text{Total number of guests who stayed} = \text{Total number of (unique) guests in the dataset}$$

<br>

That's it!

In [187]:
## Look at the unique names in the filtered records.
df["Guest"].unique()

array([nan, 'Katelyn Lindstrom', 'Erik Engström', 'Christina Cho',
       'Eric Brando', 'Adrià Crespo Aguilera', 'Nathalie Blanc',
       'Una Hopkins', 'Mark Richards', 'Clare Mackintosh', 'Le Le',
       "Paul O'Halloran", 'Dag Kjeldby', 'James Hung', 'Renato Rey Perez',
       'Alec Cohen', 'Maddy Smith', 'Lianne Aratea', 'Robert Yacucci',
       'Jordan Feucht', 'Bob Connell', 'Kelly Weir', 'Ashley Leonard',
       'Christoph Thumfart', '소민 김', 'Kari Matuseski', 'Mihai Jelea',
       'Anton Dekom', 'Melissa Gulbraa', 'Jason Visser', 'Ashley Norton',
       'Jules Vasquez', 'Lauren Provini', 'John Bashmakov',
       'Sunny Burbank', 'Katharina Volkmar', 'Madelyn Wentworth',
       'Radu Butoi', 'Sarah Franklin', 'Christina Coleman',
       'Emily Melgar', 'Taylor Eldridge', 'Will Turner', 'Will Wong',
       'Ashlie Flinn', 'Jessica Soliven', 'Christopher Rodgers',
       'Lucy Nguyen', '明月 李', 'Ella Stimson', '낙균 김', 'Jasmine Manabat',
       'Kristin Miodonski', 'Richard Gagnon',

In [188]:
## Count the number of unique names. 
## (Note: .nunique() doesn't count NaN by default, so we don't have to subtract 1 due to the NaN.)
num_guests = df["Guest"].nunique()

## Print answer.
print("Total number of guests who stayed: ", num_guests)

Total number of guests who stayed:  62


As a precautionary measure, we could further check if a guest made a reservation for a *positive* integer number of nights. Otherwise $-$ i.e., if the number of nights is 0 or `NaN` $-$ it might not make sense to say a guest has stayed, even if they have paid for the reservation somehow!


Let's see.

In [189]:
## Step 0: Check unique values in the "Nights" column. 
df["Nights"].unique()

array([nan,  3.,  2.,  4.,  5.,  1.,  7.])

It seems there are records with `NaN` (and no records with 0) in the **Nights** column.


We should *filter out* those `NaN` records:

In [190]:
stayed_df = df.loc[~pd.isna(df["Nights"])]
stayed_df

Date         Type Confirmation Code  Start Date  Nights  \
1    08/23/2019  Reservation        HMANZTA9TH  08/22/2019     3.0   
3    08/21/2019  Reservation        HMADNHAT3X  08/20/2019     2.0   
5    08/18/2019  Reservation        HMAFPFND3T  08/17/2019     3.0   
7    08/12/2019  Reservation        HMAQWFR3N5  08/11/2019     3.0   
9    08/08/2019  Reservation        HMAFT2DN4Y  08/07/2019     4.0   
..          ...          ...               ...         ...     ...   
115  01/20/2019  Reservation        HMJMMP4XZY  01/18/2019     7.0   
117  01/16/2019  Reservation        HMHC2WT2SN  01/14/2019     4.0   
119  01/12/2019  Reservation        HMS3XFNTC9  01/10/2019     3.0   
121  01/10/2019  Reservation        HMEY8DKYEN  01/08/2019     2.0   
123  01/08/2019  Reservation        HM2HMWYKXZ  01/06/2019     2.0   

                     Guest                                            Listing  \
1        Katelyn Lindstrom  Modern Gem with Upscale Finishes in San Francisco   
3            Erik Engström  Modern Gem with Upscale Finishes in San Francisco   
5            Christina Cho  Modern Gem with Upscale Finishes in San Francisco   
7              Eric Brando              Charming Family Home in San Francisco   
9    Adrià Crespo Aguilera              Charming Family Home in San Francisco   
..                     ...                                                ...   
115             Lanee Hood  Modern Gem with Upscale Finishes in San Francisco   
117        Barbera Chouaki  Modern Gem with Upscale Finishes in San Francisco   
119                 D Ayde  Modern Gem with Upscale Finishes in San Francisco   
121         Shahab Kashani  Modern Gem with Upscale Finishes in San Francisco   
123          Jerome Pascua  Modern Gem with Upscale Finishes in San Francisco   

    Details  Reference Currency   Amount  Paid Out  Host Fee  Cleaning Fee  
1       NaN        NaN      USD   446.20       NaN     13.80          70.0  
3       NaN        NaN      USD   310.40       NaN      9.60          70.0  
5       NaN        NaN      USD   431.65       NaN     13.35          70.0  
7       NaN        NaN      USD   952.54       NaN     29.46         145.0  
9       NaN        NaN      USD  1223.17       NaN     37.83         145.0  
..      ...        ...      ...      ...       ...       ...           ...  
115     NaN        NaN      USD   667.36       NaN     20.64          65.0  
117     NaN        NaN      USD   408.37       NaN     12.63          65.0  
119     NaN        NaN      USD   341.44       NaN     10.56          65.0  
121     NaN        NaN      USD   235.71       NaN      7.29          65.0  
123     NaN        NaN      USD   245.41       NaN      7.59          65.0  

[62 rows x 14 columns]

In [191]:
## Look at the unique names in the filtered records.
stayed_df["Guest"].unique()

array(['Katelyn Lindstrom', 'Erik Engström', 'Christina Cho',
       'Eric Brando', 'Adrià Crespo Aguilera', 'Nathalie Blanc',
       'Una Hopkins', 'Mark Richards', 'Clare Mackintosh', 'Le Le',
       "Paul O'Halloran", 'Dag Kjeldby', 'James Hung', 'Renato Rey Perez',
       'Alec Cohen', 'Maddy Smith', 'Lianne Aratea', 'Robert Yacucci',
       'Jordan Feucht', 'Bob Connell', 'Kelly Weir', 'Ashley Leonard',
       'Christoph Thumfart', '소민 김', 'Kari Matuseski', 'Mihai Jelea',
       'Anton Dekom', 'Melissa Gulbraa', 'Jason Visser', 'Ashley Norton',
       'Jules Vasquez', 'Lauren Provini', 'John Bashmakov',
       'Sunny Burbank', 'Katharina Volkmar', 'Madelyn Wentworth',
       'Radu Butoi', 'Sarah Franklin', 'Christina Coleman',
       'Emily Melgar', 'Taylor Eldridge', 'Will Turner', 'Will Wong',
       'Ashlie Flinn', 'Jessica Soliven', 'Christopher Rodgers',
       'Lucy Nguyen', '明月 李', 'Ella Stimson', '낙균 김', 'Jasmine Manabat',
       'Kristin Miodonski', 'Richard Gagnon', 'Ros

In [192]:
## Count the number of unique names.
num_guests = stayed_df["Guest"].nunique()

## Print answer.
print("Total number of guests who stayed: ", num_guests)

Total number of guests who stayed:  62


Hmm. It seems taking into account the number of nights does not make a difference to the result. Great!

-----

<a id='section2'></a>
### Question 2 

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        How many unique property listings are there in this file?
    </span>
</div>

In [193]:
## Step 1: Look at the unique property listings. Seems like there is an 'nan'.
df['Listing'].unique()

array([nan, 'Modern Gem with Upscale Finishes in San Francisco',
       'Charming Family Home in San Francisco',
       'Charming San Francisco Edwardian w/ Private Suite'], dtype=object)

In [194]:
## Step 2: Get the number of unique listings. (Remember: the NaN is not counted by default.)
num_listings = df['Listing'].nunique()

## Print answer.
print("Number of unique property listings: ", num_listings)

Number of unique property listings:  3


-----

<a id='section3'></a>
### Question 3

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        Compute the sum total of nights stayed per listing.
    </span>
</div>

Recall from the previous lab that `NaN` values are ignored when summing a `pd.Series`. Hence, we can just group the data by `Listing` and sum the `Nights` values for each unique `Listing`.

In [195]:
df.groupby(['Listing'])\
  .agg(total_nights_stayed = ('Nights', 'sum'))

total_nights_stayed
Listing                                                               
Charming Family Home in San Francisco                             36.0
Charming San Francisco Edwardian w/ Private Suite                  4.0
Modern Gem with Upscale Finishes in San Francisco                152.0

-----

<a id='section4'></a>
### Question 4

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        Compute the sum total of revenue generated per listing. Try this: per listing per year.
    </span>
</div>

There is no explicit column for revenue, so we'll have to figure out what 'revenue' means and how to derive it from the data in the first place.


For a start, we can safely assume that the revenue would have to be calculated from one or more of the only five columns in the dataset that are *monetary* in nature $-$ <b>Currency</b>$\hspace{0.5mm}^1$, <b>Amount</b>, <b>Host Fee</b>$\hspace{0.5mm}^2$, <b>Cleaning Fee</b>$\hspace{0.5mm}^3$, and <b>Paid Out</b>$\hspace{0.5mm}^4$. 

> $^1$ Let's check the available currencies in the data (and pray that no conversions are necessary). 

In [196]:
df['Currency'].value_counts()

USD    124
Name: Currency, dtype: int64

> It seems lady luck is on our side. The USD is the only currency in the dataset, so we won't need to do any conversions!

> $^2$ This is a service fee that a host must pay Airbnb (see [here](https://www.airbnb.com/help/article/1857/airbnb-service-fees)). It seems factored in the reservation fee reflected in the **Amount** column already.

> $^3$ This is a fee that guests pay their host (see [here](https://www.airbnb.com/help/article/2812/cleaning-fees)). It seems to be included already in the reservation fee reflected in the **Amount** column.

> $^4$ This seems to be the final payout that the host receives from Airbnb. According to [an Airbnb article](https://www.airbnb.com/help/article/459/calculating-your-payout), 
> 
> $$\text{A host's received payout} = \text{The host's revenue (including cleaning and guest fees)} - \text{Service fees (including host fees)}$$

<br>

Further, based on how [an Airbnb article](https://www.airbnb.com/help/article/2716/accessing-earnings-data#:~:text=Total%20nightly%20revenue.) defines 'total nightly revenue', it would be reasonable to posit that 'revenue' here refers to the total amount of money the host earned from reservations and other charged fees, before Airbnb's service fees, taxes, or any other fees are deducted.


Thus, to compute revenue we need to add the <u>positive</u> values in the **Amount** column and the values in the **Host Fee** column together:

In [206]:
## Step 1. Filter in positive values in Amount column, and extract only required columns.
revenue_df = df.loc[df['Amount'] > 0][['Listing', 'Date', 'Nights', 'Amount', 'Host Fee']]


> We should also remember that one of the four interesting records we [came across](1st_interesting_record) earlier has a positive value in the **Amount** column (so it will pass through the filter and be in `revenue_df`) but `NaN`s in the **Host Fee**, **Listing**, and **Nights** columns. 

In [210]:
## The special record.
revenue_df[revenue_df.isna().any(axis=1)]

Listing        Date  Nights  Amount  Host Fee
29     NaN  06/22/2019     NaN   145.5       NaN

In [209]:
## The group containing the special record.
revenue_df.loc[29:31]

Listing        Date  Nights  \
29                                                NaN  06/22/2019     NaN   
30  Modern Gem with Upscale Finishes in San Francisco  06/22/2019     2.0   
31  Modern Gem with Upscale Finishes in San Francisco  06/22/2019     1.0   

    Amount  Host Fee  
29  145.50       NaN  
30  115.43      1.38  
31  213.40      6.60

> Except for the `NaN` in the **Nights** column, the other `NaN`s will need to be replaced with appropriate values so that quantities such as total revenue by listing and revenue per night can be calculated correctly and in a bug-free manner later.

In [211]:
## Step 2(a). For that special record; replace NaN in 'Host Fee' column with 0. 
revenue_df['Host Fee'].fillna(value=0, inplace=True)

## Step 2(b). For that special record; replace NaN in 'Listing' column with the listing of its associated group.
revenue_df['Listing'].fillna(method='backfill', inplace=True)

## Step 3. Add 'Amount' and 'Host Fee' columns to get revenues (and create a 'Revenue' column for this).
revenue_df['Revenue'] = revenue_df['Amount'] + revenue_df['Host Fee']

In [213]:
## Check if above steps worked as expected.
revenue_df.loc[29:31]

Listing        Date  Nights  \
29  Modern Gem with Upscale Finishes in San Francisco  06/22/2019     NaN   
30  Modern Gem with Upscale Finishes in San Francisco  06/22/2019     2.0   
31  Modern Gem with Upscale Finishes in San Francisco  06/22/2019     1.0   

    Amount  Host Fee  Revenue  
29  145.50      0.00   145.50  
30  115.43      1.38   116.81  
31  213.40      6.60   220.00

Now that we have a separate **Revenue** column, we can address the question directly:

<u>Sum total of revenue generated per listing</u>

In [214]:
## Sum total of revenue generated per listing.
revenue_df.groupby(['Listing'])\
          .agg(total_revenue = ('Revenue', 'sum'))

total_revenue
Listing                                                         
Charming Family Home in San Francisco                   11394.00
Charming San Francisco Edwardian w/ Private Suite        1796.00
Modern Gem with Upscale Finishes in San Francisco       18769.07

---

<u>Sum total of revenue generated per listing AND per year</u>

We know, from [the EDA earlier](#date_range), that all of the records in the data fall in the year 2019. So, we should not expect any difference in the aggregated revenue values here.


Nonetheless, we could try grouping by year as well just to get more practice with `pandas`!


In order to also group by year, we will need to create a `Year` column from the existing `Date` column. 


First, let's check what the `Date` column looks like:

In [215]:
revenue_df['Date']

1      08/23/2019
3      08/21/2019
5      08/18/2019
7      08/12/2019
9      08/08/2019
          ...    
115    01/20/2019
117    01/16/2019
119    01/12/2019
121    01/10/2019
123    01/08/2019
Name: Date, Length: 63, dtype: object

It seems like the `dtype` is `object`, which suggests the dates are currently formatted as strings. 


There are many ways to extract the year from a date string. Here's one way:

In [216]:
## Convert date strings to datetime format.
revenue_df['Date'] = pd.to_datetime(revenue_df['Date'], infer_datetime_format=True)

## Check if conversion worked.
revenue_df['Date']

1     2019-08-23
3     2019-08-21
5     2019-08-18
7     2019-08-12
9     2019-08-08
         ...    
115   2019-01-20
117   2019-01-16
119   2019-01-12
121   2019-01-10
123   2019-01-08
Name: Date, Length: 63, dtype: datetime64[ns]

In [217]:
## Extract year from Date column and create a Year column.
revenue_df['Year'] = revenue_df['Date'].dt.year

## Check if the above worked.
revenue_df['Year']

1      2019
3      2019
5      2019
7      2019
9      2019
       ... 
115    2019
117    2019
119    2019
121    2019
123    2019
Name: Year, Length: 63, dtype: int64

In [218]:
## Sum total of revenue generated per listing AND per year.
revenue_df.groupby(['Listing', 'Year'])\
          .agg(total_revenue = ('Revenue', 'sum'))

total_revenue
Listing                                           Year               
Charming Family Home in San Francisco             2019       11394.00
Charming San Francisco Edwardian w/ Private Suite 2019        1796.00
Modern Gem with Upscale Finishes in San Francisco 2019       18769.07

-----

<a id='section5'></a>
### Question 5

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        Compute the mean revenue generated per night per listing.
    </span>
</div>

There are two acceptable ways of doing this that might produce different results:

1. For every listing, compute $\dfrac{\text{sum of revenue}}{\text{sum of nights}}$.


2. For every listing, compute mean of $\dfrac{\text{revenue}}{\text{night}}$.

<u>First Way</u>

In [219]:
revenue_df.groupby(['Listing'])\
          .apply(lambda x: x['Revenue'].sum()/x['Nights'].sum())

Listing
Charming Family Home in San Francisco                316.500000
Charming San Francisco Edwardian w/ Private Suite    449.000000
Modern Gem with Upscale Finishes in San Francisco    123.480724
dtype: float64

<u>Second Way</u>

In [220]:
## Method 1
revenue_df.groupby(['Listing'])\
          .apply(lambda x: (x['Revenue']/x['Nights']).mean())

Listing
Charming Family Home in San Francisco                320.175000
Charming San Francisco Edwardian w/ Private Suite    449.000000
Modern Gem with Upscale Finishes in San Francisco    127.166596
dtype: float64

In [221]:
## Method 2
revenue_df['nightly_revenue'] = revenue_df['Revenue']/revenue_df['Nights']

revenue_df.groupby(['Listing'])\
          .agg(mean_nightly_revenue = ('nightly_revenue', 'mean'))

mean_nightly_revenue
Listing                                                                
Charming Family Home in San Francisco                        320.175000
Charming San Francisco Edwardian w/ Private Suite            449.000000
Modern Gem with Upscale Finishes in San Francisco            127.166596

-----

<a id='section6'></a>
### Question 6

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        Compute the mean length of stay per listing.
    </span>
</div>

In [222]:
df.groupby(['Listing'])\
  .agg(mean_length_of_stay = ('Nights', 'mean'))

mean_length_of_stay
Listing                                                               
Charming Family Home in San Francisco                         3.600000
Charming San Francisco Edwardian w/ Private Suite             4.000000
Modern Gem with Upscale Finishes in San Francisco             2.980392

**Note:** The `NaN` values in the **Nights** column are ignored by default when computing the mean. (See [the documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#:~:text=exclude%20Na%20values.).)

-----

<a id='section7'></a>
### Question 7

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        How many stays were shorter than 3 days?
    </span>
</div>

In [223]:
answer = len(df.loc[df["Nights"] < 3])
print(answer)

21


-----

<a id='section8'></a>
### Question 8

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        How many stays exceeded 5 days?
    </span>
</div>

In [224]:
answer = len(df.loc[df["Nights"] > 5])
print(answer)

1


-----

<a id='section9'></a>
### Question 9

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        How many stays included one weekend day?
    </span>
</div>

We want to find those stays where the *window* of stay included at least one weekend day. To do so, we need to use both the **Start Date** and **Nights** columns to infer end dates (which we can then store in an **End Date** column).


Prima facie, we could check whether a stay included one weekend day via the following two steps:

1. Map the dates in the **Start Date** and **End Date** columns to numerical days of the week. 

> So: Monday will be mapped to 0, Tuesday will be mapped to 1, ..., and Sunday will be mapped to 6. See the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.dayofweek.html#pandas.Series.dt.dayofweek).

2. Count the number of stays where Saturday (5) or Sunday (6) lies in the range [start day, end day]. In terms of code, we could check whether 5 or 6 $\geq$ start day AND 5 or 6 $\leq$ end day.

> For example, if the start date and end date for a stay were Monday (0) and Wednesday (2), we could simply check that neither 5 nor 6 is within [0, 2] to infer that this stay did not include a weekend day.

As you might have anticipated already, this method is flawed because it doesn't work for stays that cross over weeks! 


Let's say we have a stay with start and end dates of Friday (4) and Tuesday (1), respectively. Clearly, this stay did include a weekend day (two, in fact). But the above method would not indicate so, since 5 or 6 $\not\leq$ 1. How should we go about this, then?

.
.
.

There is, in fact, a better way that works. Instead of inferring the end dates (and converting them to numerical days of the week), we can simply add the number of nights stayed (**Nights** column) to the numerical start date (**Start Date** column, converted to numerical days of the week). Let's call this resultant sum $X$. If $X > 4$, then we know the stay included at least one weekend day. Otherwise, the stay did not include any weekend day. 


Brilliant, isn't it?

---
Let's start by converting the date strings in the 'Start Date' column to `datetime` format:

In [225]:
## Convert date strings to datetime format.
df['StayStartDate'] = pd.to_datetime(df['Start Date'], infer_datetime_format=True)

In [226]:
## Check if conversion worked.
print("Start Date:")
print(df['StayStartDate'])

Start Date:
0            NaT
1     2019-08-22
2            NaT
3     2019-08-20
4            NaT
         ...    
119   2019-01-10
120          NaT
121   2019-01-08
122          NaT
123   2019-01-06
Name: StayStartDate, Length: 124, dtype: datetime64[ns]


---
Next, let's convert the dates to numerical days of week (0 to 6):

> **REMEMBER:** Monday is mapped to 0, Tuesday is mapped to 1, ..., and Sunday is mapped to 6. It doesn't go from 1 to 7!

In [227]:
# Convert each date to a day of week (number from 0 to 6).
df['StayStartDate'] = df['StayStartDate'].dt.dayofweek

In [228]:
## Check if conversions worked.
print("Start Date (Day of Week):")
print(df['StayStartDate'])

Start Date (Day of Week):
0      NaN
1      3.0
2      NaN
3      1.0
4      NaN
      ... 
119    3.0
120    NaN
121    1.0
122    NaN
123    6.0
Name: StayStartDate, Length: 124, dtype: float64


---

Now, let's add the **StayStartDate** and **Nights** columns together to get the variable we desire (and assign the sums to a column called **X**):

In [229]:
df['X'] = df['StayStartDate'] + df['Nights']

---

Then, all we need to do to complete the problem is count the number of records where $X > 4$.

In [230]:
## Check the filtered records.
weekend_df = df.loc[df['X'] > 4]
weekend_df

Date         Type Confirmation Code  Start Date  Nights  \
1    08/23/2019  Reservation        HMANZTA9TH  08/22/2019     3.0   
5    08/18/2019  Reservation        HMAFPFND3T  08/17/2019     3.0   
7    08/12/2019  Reservation        HMAQWFR3N5  08/11/2019     3.0   
9    08/08/2019  Reservation        HMAFT2DN4Y  08/07/2019     4.0   
13   08/03/2019  Reservation        HMJR8RN3WD  08/02/2019     3.0   
15   07/28/2019  Reservation        HM4228XARZ  07/27/2019     5.0   
17   07/25/2019  Reservation        HMAJJF2Y2D  07/24/2019     3.0   
19   07/18/2019  Reservation        HMAEFBHA8R  07/17/2019     4.0   
21   07/14/2019  Reservation        HMABCPWC5Q  07/13/2019     4.0   
23   07/10/2019  Reservation        HMAKP43DQS  07/09/2019     4.0   
25   07/04/2019  Reservation        HMAKXATN3A  07/03/2019     4.0   
27   06/27/2019  Reservation        HMSZQWB83R  06/26/2019     4.0   
30   06/22/2019  Reservation        HMAAEB8BRK  06/21/2019     2.0   
31   06/22/2019  Reservation        HMAQW2PHXY  06/21/2019     1.0   
35   06/15/2019  Reservation        HMAN4N55ZJ  06/14/2019     4.0   
39   06/10/2019  Reservation        HMANJCY8PQ  06/09/2019     3.0   
41   06/06/2019  Reservation        HMZMDEAJWP  06/05/2019     3.0   
43   05/29/2019  Reservation        HMBDZ2PBFC  05/28/2019     4.0   
44   06/02/2019  Reservation        HMW4EBB2RS  06/01/2019     3.0   
46   05/26/2019  Reservation        HMAFC4KRTR  05/25/2019     3.0   
48   05/24/2019  Reservation        HMA8XTFYSD  05/23/2019     2.0   
50   05/20/2019  Reservation        HMACSQ9EYH  05/19/2019     4.0   
52   05/18/2019  Reservation        HMAHP5WNE2  05/17/2019     2.0   
54   05/13/2019  Reservation        HMAMMD42XP  05/12/2019     4.0   
56   05/08/2019  Reservation        HMPJ5BSMWE  05/07/2019     5.0   
58   05/06/2019  Reservation        HMACEWW5C8  05/05/2019     2.0   
60   05/04/2019  Reservation        HMAEARM8CP  05/03/2019     2.0   
63   05/01/2019  Reservation        HMYQNM2ECW  04/30/2019     5.0   
66   04/27/2019  Reservation        HMARRKJPA8  04/26/2019     4.0   
71   04/22/2019  Reservation        HMAKS82CT4  04/21/2019     1.0   
73   04/13/2019  Reservation        HMCF5FDBWK  04/12/2019     2.0   
75   04/07/2019  Reservation        HMSB84HSJQ  04/06/2019     5.0   
77   04/05/2019  Reservation        HMABZ95YJH  04/04/2019     2.0   
81   03/29/2019  Reservation        HM5K3R8M42  03/28/2019     2.0   
83   03/17/2019  Reservation        HMAJPKNQB8  03/16/2019     4.0   
85   03/15/2019  Reservation        HM4ZB3HFPA  03/14/2019     2.0   
87   03/12/2019  Reservation        HMAHD9QBH3  03/10/2019     4.0   
89   03/09/2019  Reservation        HMXTE3F2SB  03/07/2019     3.0   
94   03/02/2019  Reservation        HMEEP8EA9D  02/28/2019     2.0   
95   03/03/2019  Reservation        HMXHTWYYS8  03/01/2019     4.0   
97   02/25/2019  Reservation        HMX4YQNFN9  02/23/2019     3.0   
101  02/18/2019  Reservation        HMK2MTRSH2  02/16/2019     2.0   
105  02/10/2019  Reservation        HMKSJXYXSR  02/08/2019     5.0   
109  02/05/2019  Reservation        HM4SJZMRSP  02/03/2019     3.0   
111  02/02/2019  Reservation        HMP2FZA9NS  01/31/2019     3.0   
113  01/27/2019  Reservation        HMJZZ4AQ33  01/25/2019     2.0   
115  01/20/2019  Reservation        HMJMMP4XZY  01/18/2019     7.0   
119  01/12/2019  Reservation        HMS3XFNTC9  01/10/2019     3.0   
123  01/08/2019  Reservation        HM2HMWYKXZ  01/06/2019     2.0   

                     Guest                                            Listing  \
1        Katelyn Lindstrom  Modern Gem with Upscale Finishes in San Francisco   
5            Christina Cho  Modern Gem with Upscale Finishes in San Francisco   
7              Eric Brando              Charming Family Home in San Francisco   
9    Adrià Crespo Aguilera              Charming Family Home in San Francisco   
13             Una Hopkins              Charming Family Home in San Francisco   
15           Mark Richar

In [231]:
## Compute the answer.
answer = len(weekend_df)
print("Number of stays that included one weekend day: ", answer)

Number of stays that included one weekend day:  49


-----

<a id='section10'></a>
### Question 10

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        How many stays began on a Tuesday?
    </span>
</div>

In [232]:
## Filter in those stays that began on a Tuesday. (NOTE: Tuesday = 1.)
tuesday_df = df.loc[df['StayStartDate'] == 1]
tuesday_df

Date         Type Confirmation Code  Start Date  Nights  \
3    08/21/2019  Reservation        HMADNHAT3X  08/20/2019     2.0   
23   07/10/2019  Reservation        HMAKP43DQS  07/09/2019     4.0   
33   06/19/2019  Reservation        HMANEJTK4E  06/18/2019     3.0   
43   05/29/2019  Reservation        HMBDZ2PBFC  05/28/2019     4.0   
56   05/08/2019  Reservation        HMPJ5BSMWE  05/07/2019     5.0   
63   05/01/2019  Reservation        HMYQNM2ECW  04/30/2019     5.0   
64   05/01/2019  Reservation        HMAKW2PEZW  04/30/2019     3.0   
91   03/07/2019  Reservation        HMMWNJPF4B  03/05/2019     2.0   
99   02/21/2019  Reservation        HM4KMJSHB2  02/19/2019     3.0   
121  01/10/2019  Reservation        HMEY8DKYEN  01/08/2019     2.0   

                   Guest                                            Listing  \
3          Erik Engström  Modern Gem with Upscale Finishes in San Francisco   
23           Dag Kjeldby              Charming Family Home in San Francisco   
33         Lianne Aratea  Modern Gem with Upscale Finishes in San Francisco   
43        Ashley Leonard  Modern Gem with Upscale Finishes in San Francisco   
56          Jason Visser  Modern Gem with Upscale Finishes in San Francisco   
63        Lauren Provini  Modern Gem with Upscale Finishes in San Francisco   
64        John Bashmakov  Modern Gem with Upscale Finishes in San Francisco   
91   Christopher Rodgers  Modern Gem with Upscale Finishes in San Francisco   
99                  낙균 김  Modern Gem with Upscale Finishes in San Francisco   
121       Shahab Kashani  Modern Gem with Upscale Finishes in San Francisco   

    Details  Reference Currency   Amount  Paid Out  Host Fee  Cleaning Fee  \
3       NaN        NaN      USD   310.40       NaN      9.60          70.0   
23      NaN        NaN      USD  1223.17       NaN     37.83         145.0   
33      NaN        NaN      USD   385.09       NaN     11.91          70.0   
43      NaN        NaN      USD   485.97       NaN     15.03          65.0   
56      NaN        NaN      USD   591.70       NaN     18.30          65.0   
63      NaN        NaN      USD   274.02       NaN      3.80           0.0   
64      NaN        NaN      USD   385.09       NaN     11.91          70.0   
91      NaN        NaN      USD   255.11       NaN      7.89          65.0   
99      NaN        NaN      USD   351.14       NaN     10.86          65.0   
121     NaN        NaN      USD   235.71       NaN      7.29          65.0   

     StayStartDate    X  
3              1.0  3.0  
23             1.0  5.0  
33             1.0  4.0  
43             1.0  5.0  
56             1.0  6.0  
63             1.0  6.0  
64             1.0  4.0  
91             1.0  3.0  
99             1.0  4.0  
121            1.0  3.0

In [233]:
## Get answer.
answer = len(tuesday_df)
print("Number of stays that began on Tuesday: ", answer)

Number of stays that began on Tuesday:  10


-----

<a id='section11'></a>
### Question 11

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        The field <b>Host Fee</b> is how much the Airbnb operator paid to Airbnb as a commission. What is the sum total commission paid to Airbnb by listing?
    </span>
</div>

In [234]:
df.groupby(['Listing'])\
  .agg(total_Airbnb_commission = ('Host Fee', 'sum'))

total_Airbnb_commission
Listing                                                                   
Charming Family Home in San Francisco                               341.82
Charming San Francisco Edwardian w/ Private Suite                    53.88
Modern Gem with Upscale Finishes in San Francisco                   547.63

-----

<a id='section12'></a>
### Question 12

<div class="alert alert-block alert-warning"> 
    <span style="color:black">
        Airbnb is run like a business, on which profit (revenue minus expenses) is taxed. Let's assume a tax rate of 33%. What is the sum total of taxes paid to Airbnb?
    </span>
</div>

If you read through the [EDA section](#section0_2) carefully, it would become clear that the values in the **Paid Out** column already have the revenue (e.g., reservation fees, cleaning fees, etc.) and expenses (e.g., host fees, resolution adjustments, photography payment, etc.) factored in. 

> After all, this is the money that Airbnb transferred to the host $-$ it would not make sense for Airbnb to transfer the revenue *before* subtracting the expenses.


Thus, the values in the **Paid Out** column *are* the profits!

.
.
.

Therefore, all we need to do to obtain the sum total of taxes that the host paid to Airbnb is:


> Multiply the sum of the values in the **Paid Out** column by 0.33.

That's it!

In [235]:
sum_taxes = 0.33*df['Paid Out'].sum()
sum_taxes = round(sum_taxes, 2)
print(f"Sum total of taxes paid to Airbnb: {sum_taxes} USD")

Sum total of taxes paid to Airbnb: 10153.68 USD


And we are done. Voilà!

<div class="alert alert-block alert-success">
    
<b><center> End of Solutions <b>